composer/trainer/trainer.py

```diff
@@ -2670,12 +2677,29 @@ def _eval_loop(
                 metric.reset()
 
             dataloader = self.state.dataloader
+            dist_sampler = None
+            drop_last = None
+            dataset_len = None
+            last_batch = False
             if isinstance(dataloader, DataLoader) and isinstance(dataloader.sampler, DistributedSampler):
                 # The distributed sampler uses `set_epoch` to set the random seed
                 # Because evaluation can run on each batch, we use the batch to seed the sampler
                 # so each evaluation will get a proper shuffle.
                 # The epoch provided to `set_epoch` need not be sequential, so this is fine.
-                dataloader.sampler.set_epoch(int(self.state.timestamp.batch))
+                dist_sampler = dataloader.sampler
+                dist_sampler.set_epoch(int(self.state.timestamp.batch))
+                drop_last = dataloader.drop_last
+                # Only compute the dataset length if drop_last is False, as otherwise we don't need
+                # to remove any duplicate samples.
+                if drop_last == False:
+                    try:
+                        dataset_len = len(dist_sampler.dataset)  # type: ignore
+                    except AttributeError:
+                        warnings.warn("DistributedSampler's dataset attribute does not have length. When "
+                                      '`drop_last=False`, metrics may be incorrect as DistributedSampler '
+                                      'duplicates samples to make all batches the same size. To fix this, '
```

Comment:
correct me if im wrong, but it doesn't make "all batches the same size", but rather, "the dataset size divisible by the world size", right?